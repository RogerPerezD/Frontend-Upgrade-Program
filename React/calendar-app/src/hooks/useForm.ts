import { FormEvent, useState } from "react";


export const useForm = <T>( initialState: T ): [ T, (e: FormEvent<HTMLInputElement>)=>void, ()=>void ] => {
    const [values, setValues] = useState<T>(initialState);

    const handleInput = ( { currentTarget }: FormEvent<HTMLInputElement> ): void =>{
        setValues({
            ...values,
            [ currentTarget.name ]: currentTarget.value
        });
    }

    const reset  = () =>{
        setValues( initialState );
    }

    return [ values, handleInput, reset ];
};