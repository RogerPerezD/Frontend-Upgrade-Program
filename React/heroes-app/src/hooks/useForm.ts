import { FormEvent, useState } from "react";

type FieldsForm = {
    searchText: string
}

export const useForm = ( initialState: FieldsForm ): [FieldsForm, (e: FormEvent<HTMLInputElement>)=>void, ()=>void ] => {
    const [values, setValues] = useState<FieldsForm>(initialState);

    const handleInput = ( { currentTarget }: FormEvent<HTMLInputElement> ): void =>{
        setValues({
            ...values,
            [ currentTarget.name ]: currentTarget.value
        });
    }

    const reset  = () =>{
        setValues( initialState );
    }

    return [ values, handleInput, reset ];
};
