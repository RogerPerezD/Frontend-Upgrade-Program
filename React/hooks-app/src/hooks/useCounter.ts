import { useState } from "react";



const useCounter = ( initialState: number = 0) => {
  
    const [state, setState] = useState(initialState);

    const increment = ( factor: number = 1) =>{
        setState( state + factor);
    }

    const decrement = ( factor: number = 1) =>{
        setState( state - factor);
    }

    const reset = () =>{
        setState( initialState );
    }

    return {
        state,
        increment,
        decrement,
        reset
    };
};

export default useCounter;
