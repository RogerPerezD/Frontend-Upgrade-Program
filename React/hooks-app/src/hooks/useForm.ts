import { FormEvent, useState } from "react";
import { FieldsForm } from "../components/02-useEffect/FormWithCustomHook";


export const useForm = ( initialState: FieldsForm ): [FieldsForm, (e: FormEvent<HTMLInputElement>)=>void, ()=>void ] => {
    const [values, setValues] = useState<FieldsForm>(initialState);

    const handleInput = ( { currentTarget }: FormEvent<HTMLInputElement> ): void =>{
        setValues({
            ...values,
            [ currentTarget.name ]: currentTarget.value
        });
    }

    const reset  = () =>{
        setValues( initialState );
    }

    return [ values, handleInput, reset ];
};
