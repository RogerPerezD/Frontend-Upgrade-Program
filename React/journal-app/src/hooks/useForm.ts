import { FormEvent, useState } from 'react';

type FieldsForm = {
    email: string,
    password: string,
    name?: string,
    password2?: string
} 

export const useForm = ( initialState = {} as FieldsForm ): [FieldsForm, (e: FormEvent<HTMLInputElement>)=>void, ()=>void ] => {
    
    const [values, setValues] = useState<FieldsForm>(initialState);

    const reset = () => {
        setValues( initialState );
    }


    const handleInputChange = ({ currentTarget  }: FormEvent<HTMLInputElement>) => {

        setValues({
            ...values,
            [ currentTarget.name ]: currentTarget.value
        });

    }

    return [ values, handleInputChange, reset ];
}