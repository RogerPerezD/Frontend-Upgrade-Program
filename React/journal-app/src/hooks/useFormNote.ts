import { FormEvent, useState } from 'react';

type FieldsFormNote = {
    title: string,
    body: string
} 

export const useFormNote = ( initialState = {} as FieldsFormNote ): [FieldsFormNote, (e: FormEvent<(HTMLInputElement | HTMLTextAreaElement)>)=>void, (defectValue : FieldsFormNote)=>void ] => {
    
    const [values, setValues] = useState<FieldsFormNote>(initialState);

    const reset = ( defectValue : FieldsFormNote = initialState) => {
        setValues( defectValue );
    }


    const handleInputChange = ({ currentTarget  }: FormEvent<(HTMLInputElement | HTMLTextAreaElement)>) => {

        setValues({
            ...values,
            [ currentTarget.name ]: currentTarget.value
        });

    }

    return [ values, handleInputChange, reset ];
}