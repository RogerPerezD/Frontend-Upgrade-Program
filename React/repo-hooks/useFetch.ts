import { useState,useEffect, useRef } from 'react';

type fecthState = {
    data: any,
    loading: boolean,
    error: any
}

export const useFetch = ( url: string) => {
  
    const [state, setState] = useState<fecthState>({
        data: null,
        loading: true,
        error: null
    }); 

    // variable que nos ayudara a manejar si el componente
    // esta montado o no 
    const isMounted = useRef<boolean>(true);

    useEffect(() => {
      return () => {
          isMounted.current = false;
      };
    }, []);
    

    useEffect(() => {
        // Esto para cada vez que cargemos un nuevo quote se vea el mensja de loading
        setState( {data: null, loading: true,error: null});

        fetch( url )
        .then( ( resp ) => resp.json() )
        .then( ( data ) => {

                if (isMounted.current) {
                    setState( {
                        data,
                        loading: false,
                        error: null
                    });
                }
        })
        .catch( () =>{
            setState( {
                data: null,
                loading: false,
                error: 'Fatal error'
            });
        });
    }, [url]);
    
    return state;
};
