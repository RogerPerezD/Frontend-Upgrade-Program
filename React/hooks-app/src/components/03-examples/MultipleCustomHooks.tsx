import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import useCounter from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

const { state: counter, increment } = useCounter(1);

const url: string = `https://www.breakingbadapi.com/api/quotes/${counter}`;

const { loading, data } = useFetch( url );
const { author, quote } = !!data && data[0];
// console.log(author, quote);

// const handleClick = () =>{

// }

  return (
  <div>
      <h1>BreakingBad Quotes</h1> 
      <hr />

      {
          loading ? 
          (
            <div className="alert alert-info text-center">
                Loading............
            </div>
          ) :
          (
            <blockquote className="blockquote text-end">
                <p className="mb-3">
                    { quote }
                </p>
                <footer className="blockquote-footer"> { author } </footer>
            </blockquote>
          )
      }

      <button className='btn btn-primary'
      onClick={ increment }> Add next Quote </button>
  </div>
  );
};
