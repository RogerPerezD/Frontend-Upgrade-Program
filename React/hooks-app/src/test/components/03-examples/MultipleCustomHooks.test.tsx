import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import useCounter from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Test in <MultipleCustomHooks/>', () => {
    const mockuseFetch= useFetch as jest.MockedFunction<typeof useFetch>;
    const mockuseCounter= useCounter as jest.MockedFunction<typeof useCounter>;

    beforeEach(()=>{
        mockuseCounter.mockReturnValue({
            state: 10,
            increment: () => {},
            decrement: ()=> {},
            reset: ()=> {}
        });
    });
    
    
    test('should show the component succesfully', () => {
    
        mockuseFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });
        
        const wrapper = shallow( <MultipleCustomHooks />);

        expect( wrapper ).toMatchSnapshot();
    });

    test('should show the information', () => {

        mockuseFetch.mockReturnValue({
            data: [{
                author: 'Roger',
                quote: 'Hello world'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks />);
        const author =  wrapper.find('footer').text().trim();
        const quote =  wrapper.find('.mb-3').text().trim();

        expect( wrapper.find('.alert').exists()).toBe( false );
        expect(quote).toBe( 'Hello world' );
        expect(author).toBe( 'Roger' );
    });

});