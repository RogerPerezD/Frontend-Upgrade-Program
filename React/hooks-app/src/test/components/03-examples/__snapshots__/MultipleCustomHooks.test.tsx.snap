// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <MultipleCustomHooks/> should show the component succesfully 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading............
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
     Add next Quote 
  </button>
</div>
`;
